app/code/Magento/InventoryExportStock/Model/ResourceModel/StockIndexDumpProcessor.php

```diff
@@ -0,0 +1,139 @@
+
+    /**
+     * GetStockIndexDump constructor
+     *
+     * @param StockIndexTableNameResolverInterface $stockIndexTableNameResolver
+     * @param ResourceConnection $resourceConnection
+     * @param ManageStockCondition $manageStockCondition
+     * @param GetQtyForNotManageStock $getQtyForNotManageStock
+     */
+    public function __construct(
+        StockIndexTableNameResolverInterface $stockIndexTableNameResolver,
+        ResourceConnection $resourceConnection,
+        ManageStockCondition $manageStockCondition,
+        GetQtyForNotManageStock $getQtyForNotManageStock
+    ) {
+        $this->stockIndexTableNameResolver = $stockIndexTableNameResolver;
+        $this->resourceConnection = $resourceConnection;
+        $this->manageStockCondition = $manageStockCondition;
+        $this->getQtyForNotManageStock = $getQtyForNotManageStock;
+    }
+
+    /**
+     * Provides sku and qty of products dumping them from stock index table
+     *
+     * @param int $stockId
+     * @return array
+     * @throws Zend_Db_Select_Exception
+     */
+    public function execute(int $stockId): array
+    {
+        $this->connection = $this->resourceConnection->getConnection();
+        $select = $this->connection->select();
+        $select->union([
+            $this->getStockItemSelect($stockId),
+            $this->getStockIndexSelect($stockId)
+        ]);
+
+        return $this->connection->fetchAll($select);
+    }
+
+    /**
+     * Provides stock select
+     *
+     * @param int $stockId
+     * @return Select
+     */
+    private function getStockIndexSelect(int $stockId): Select
+    {
+        $stockIndexTableName = $this->resourceConnection
+            ->getTableName($this->stockIndexTableNameResolver->execute($stockId));
+
+        return $this->connection->select()
+            ->from(
+                $stockIndexTableName,
+                [
+                    'qty' => 'quantity',
+                    'sku' => 'sku'
+                ]
+            );
+    }
+
+    /**
+     * Provides stock item select
+     *
+     * @param int $stockId
+     * @return Select
+     */
+    private function getStockItemSelect(int $stockId): Select
+    {
+        $legacyStockItemTable = $this->resourceConnection
+            ->getTableName('cataloginventory_stock_item');
+        $productEntityTable = $this->resourceConnection
+            ->getTableName('catalog_product_entity');
+        $select = $this->connection->select();
+
+        $select->from(
+            ['legacy_stock_item' => $legacyStockItemTable],
+            new Zend_Db_Expr('"' . $this->getQtyForNotManageStock->execute() . '" as qty')
+        )->join(
+            ['product_entity' => $productEntityTable],
+            'legacy_stock_item.product_id = product_entity.entity_id',
+            ['sku']
+        )->where(
+            $this->manageStockCondition->execute($select)
+        )->where(
+            'legacy_stock_item.stock_id = ?',
```

Comment:
this would not work, because we don't duplicate data in `legacy_stock_item` table per each stock.
So, all the data in  `legacy_stock_item` table  have Default Stock assigned.

Reply:
And how to get not in stock products with sql query? If this one doesn't work?